src/prep.py

```diff
@@ -0,0 +1,331 @@
+import argparse
+import os
+from pathlib import Path
+import re
+import shutil
+import subprocess
+import sys
+import textwrap
+from typing import Optional
+
+from icecream import ic
+
+try:
+    from .cli import (
+        error,
+        info,
+        success,
+        warn,
+        todo,
+        validate_path_from_cli,
+        validate_target_from_cli,
+        validate_build_cmd_from_cli,
+    )
+    from .types import Command
+except ImportError:
+    """
+    This is necessary to safeguard the user from running this script directly
+    which will generate errors due to relative importing rules in Python
+    """
+    print(
+        f"You should not run {__file__} as a standalone script.\n"
+        f"Use `$ python3 reduce.py {Path(__file__).stem}` instead!",
+        file=sys.stderr,
+    )
+    sys.exit(1)
+
+
+DEBUG = False
+ic.disable()
+
+
+def make_dot_i_file(
+    build_cmd: Command,
+    target: Path,
+    path_to_linux: Path,
+    output_dir: Path,
+    force_rm_existing_target: bool,
+) -> str:
+    """
+    Make the preprocessed .i file which can then be compiled sans build system
+    """
+
+    if (abs_target := path_to_linux / target).exists():
+        double_check_removal_with_user(
+            file=abs_target, force_rm=force_rm_existing_target
+        )
+
+    info(f"Making {target}...")
+    make_output = subprocess.run(
+        [*build_cmd, target], check=False, capture_output=True, cwd=path_to_linux
+    )
+    ic(make_output)
+
+    if make_output.returncode == 0:
+        info(f"Moving generated {target} file to {output_dir}")
+        shutil.move(path_to_linux / target, output_dir / target.name)
+        success(f"Successfully generated {target.name}")
+
+    return make_output.stdout.decode("utf-8")
+
+
+def get_compiler_invocation(target: Path, make_stdout: str) -> Command:
+    """
+    Parse out the CC invocation from make's raw output with V=1 on
+    """
+    pattern = f".*-o {target}.*"
+    matches = re.findall(pattern, make_stdout)
+    ic(matches)
+    assert (
+        len(matches) == 1
+    ), "Too many (or too few) compiler invocation matches!\n\
+            (not your fault, it's prepreduce's fault)"
+
+    main_invocation = matches[0].split()
+    main_invocation.insert(-1, "-c")
+    return clean_compiler_invocation(main_invocation)
+
+
+def clean_compiler_invocation(cc_invocation: Command) -> Command:
+    """
+    Remove all preprocessor flags as well as any relative pathing in invocation
+    """
+
+    to_remove = ("-I", "-D", "-Wp", "-include", "-Werror", "./", "-U", "-E")
+    rules = [(lambda rule: lambda flag: rule not in flag)(rule) for rule in to_remove]
+
+    cleaned = cc_invocation.copy()
+    for rule in rules:
+        cleaned = filter(rule, cleaned)
+
+    cleaned = list(cleaned)  # eagerly convert iterator to list
+
+    ic(cleaned)
+
+    return cleaned
+
+
+def write_test_script(
+    *,
```

Comment:
what does `*` in a parameter list mean?

Reply:
Mandatory kwargs to follow.

```py
def foo(*, a: int, b: int) -> int:
   ...

# now any invocation of foo requires kwargs
result = foo(a=4, b=6)
```

Reply:
something like 
```py
foo(3, 4) # raises TypeError: foo() takes 0 positional arguments but 2 were given
```

Reply:
I usually add `(*, ...)` if my function takes 3 or 4 very similar argumemts (like paths, or flags, or numbers) so that the caller cannot mess it up. (also helps my linter give me argument-hints when I'm typing an invocation)

Reply:
neat! TIL